[Benjamin Lim] iP

itzblim opened the pull request:
### Duke
> “The best thing about Duke is that it's the best app ever made.” - Albert Einstein ([source](https://www.goodreads.com/quotes))

This duke is:
- simple
- **really fast** at processing inputs
- ~~easy~~ *wonderful* to use

So why not,

1. Download it
2. Double-click it
3. Enjoy! :star_struck:

Here's the `main` method:
```javascript
public static void main(String[] args) {
    Application.launch(Main.class, args);
}
```

You can try:
- listing all tasks using "list"
- adding a todo task using "todo [description]"
- adding an event using "event [description] /at YYYY-MM-DD"
- adding a deadline using "deadline [description] /by YYYY-MM-DD"
- marking a task as completed using "done [task number]"
- deleting a task as completed using "delete [task number]"
- finding a task using "find [search query]"
- exiting Duke using "bye"

Commit:
Add Gradle support

Commit:
Level-1 Greet, Echo, Exit

Commit:
Level 2. Add, List

Commit:
Level 3. Mark as Done

Commit:
Level 4. ToDos, Events, Deadlines

Commit:
A-TextUiTesting: Automated Text UI Testing

Commit:
Level 5. Handle Errors

Commit:
Level 6. Delete

Commit:
A-Enums: Use Enums

Commit:
Level 7. Save

Commit:
Add date parsing and formatting

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Remove methods for retrieving time and update unit testing

Commit:
Add class for user interactions

Commit:
Create classes for storage, ui, parser and task list, and update test cache

Commit:
Update javadocs

Commit:
Divide classes into packages

Commit:
Add JUnit tests

Commit:
Added manifest for building JAR files

Commit:
Add javadoc comments, rename test methods

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Fix violations of the coding standards and Git standards required.

There are several violations of the coding standard.

A common standard followed is much better in the long run as it allows others to easily read and add to the code base.

Let's update the violations as such.

Commit:
Add a find command for users to use a keyword as a query.

There is no good way to search for specific tasks.

A find command is more user-friendly than having to search through the entire list of tasks using the list command.

Let's update the command list by implementing this function.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
Fix bugs with junit tests and dependencies

Junit tests have problems referring to the main module.

The bugs with the dependencies would render developers unable to properly test the main module.

Let's update the test module to fix the testing methods.

Commit:
Add manifest file for generation of JAR files

There is no manifest included in the Git.

Including a manifest helps with the generation of JAR executables.

Let's include the manifest file.

Commit:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

Commit:
Automate prject builds using Gradle

The project had no gradle support.

Working without gradle support could lead to inefficient testing.

Let's add gradle support to this project.

Commit:
Merge branch 'branch-A-Gradle'

Commit:
Use CheckStyle to detect coding style violations

There were minimal checks completed to ensure compliance with the coding standard.

Complying with the coding standard helps to make the code and documentation seem seamless.

Let's update the code with the changes needed to align with the coding standard.

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add a GUI to Duke.

There is no graphical interface for Duke.

Adding a graphical user interface makes it much more user friendly.

Let's add a graphical interface to improve the user experience.

Commit:
Merge branch 'branch-Level-10'

Commit:
Use var-args

There is no use of var-args in this project.

The use of varags can help to make applicable functions more flexible by allowing different numbers of arguments.

Let's update the project so that varargs is used where relevant.

Commit:
Merge branch 'branch-A-Varargs'

jeromewjj commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+        try {
+            this.byTime = LocalDate.parse(by);
+        } catch (DateTimeParseException e) {
+            throw new DukeException("Could not parse time input.");
```
Maybe you could add the required format so that user will know what to enter.

jeromewjj commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,40 @@
+    public String toString() {
+        return "[E]" + super.toString() + " (at: "
+                + atTime.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                + ")";
```
Maybe this could be moved to the previous line. It will make the code looks cleaner.

jeromewjj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+                command.substring(6, atIndex - 1),
+                command.substring(atIndex + 4)));
+        } else {
+            throw new DukeException("What on earth does that mean? Are you okay?");
```
I like this message.

jeromewjj commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,92 @@
+                        } else {
+                            exit = true;
+                        }
+                    } catch (DukeException e) {
```
Maybe you could avoid deep nesting.

jeromewjj reviewed (commented):
Overall, the code is easy to read and not much coding standard violation.

hungkhoaitay commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,67 @@
+     * @throws DukeException Thrown when an invalid command is given.
+     */
+    public String[] parseCommand(String command) throws DukeException {
+        if (command.equals("bye")) {
```
How about using `switch` :)

hungkhoaitay reviewed (commented):
The code is well written, just one suggestion that I made for yours. Well done! 👍

Commit:
Update user and Duke pictures

Default pictures are used for user and Duke.

Replacing the default pictures with custom graphics helps to differentiate this Duke program from other chatbots.

Let's
* update the picture used to represent the picture
* update the picture used to represent Duke (the chatbot)

CommanderW324 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,40 @@
+ * Class for event tasks.
+ */
+public class Event extends Task {
+    private final LocalDate atTime;
+    /**
```
I think the class variable should be in camelCase ?

CommanderW324 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,49 @@
+        todo, deadline, event
+    }
+    protected TaskType category;
+    private final String description;
+    private boolean completed;
```
Maybe this one can be name isCompleted ?

CommanderW324 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,118 @@
+        if (isInvalidID(index)) {
+            throw new DukeException("Task ID out of range!");
+        }
+        Task t = taskArray.remove(index);
+        return new String[] {
```
maybe use a more understandable name ? tasks ?

CommanderW324 reviewed (commented):
Overall, very nicely styled and readable !

Commit:
Add assertions

There are no assertions in the entire project.

Using assertions can help to document important assumptions that should hold at various points in the code.

Let's add assertions for various classes to ensure that assumptions hold true.

Commit:
Improve code quality

Some classes have suboptimal code quality, with long methods and deep nesting.

Improving the code quality can help with readability and reducing unsafe practices.

Let's make the changes to the code quality.

itzblim committed:
Merge pull request #1 from itzblim/branch-A-Assertions

Add assertions

There are no assertions in the entire project.

Using assertions can help to document important assumptions that should hold at various points in the code.

Let's add assertions for various classes to ensure that assumptions hold true.

Commit:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/MainWindow.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/Ui.java

itzblim committed:
Merge pull request #2 from itzblim/branch-A-CodeQuality

Improve code quality

Some classes have suboptimal code quality, with long methods and deep nesting.

Improving the code quality can help with readability and reducing unsafe practices.

Let's make the changes to the code quality.

Commit:
Set up Continuous Integration

There is no continuous integration for the project.

Setting up continuous integration allows for some work to be automated every time events are triggered.

Let's add continuous integration.

Commit:
Add a way to view a date's schedule

There is no way to get tasks for a specific date.

Adding a schedule command allows users to search for tasks chronologically.

Let's add a command to allow users to find tasks for a specific date.

Commit:
Merge branch 'branch-B-ViewSchedules'

Commit:
Improve the GUI

The GUI uses the same colors for the user and duke. In addition, the pictures used look unprofessional.

Improving the GUI to make it more polished helps users to differentiate between user and duke messages and become more satisfied by the aesthetics.

Let's improve the GUI with better colors and images.

Commit:
Add a user guide

There is no guidance for users as to how to run the chatbot.

Adding a user guide helps to allow users to figure out what commands are available and how to use them.

Let's add the user guide.

itzblim committed:
Set theme jekyll-theme-slate

itzblim committed:
Set theme jekyll-theme-leap-day

Commit:
Update user guide.

The user guide in the docs folder is not updated.

Failing to update the user guide in the docs folder means the website displays no information.

Let's update the user guide in the docs folder.

Commit:
Update code to match code quality standards

There are parts of the code which do not adhere to the quality standards required.

Matching code quality standards ensures readability and consistency.

Let's update the code to reach the standards set.

Commit:
Update Junit tests

The class for encapsulating task lists is not being tested.

Adding tests for more classes helps to decrease the possibility of undiscovered bugs.

Let's update the Junit tests as such.
